src/storage/mod.rs

```diff
@@ -189,6 +189,120 @@ where
     }
 }
 
+/// An entry to a storage which has a component associated to the entity.
+pub struct OccupiedEntry<'a, 'b: 'a, T: 'b, D: 'b> {
+    entity: Entity,
+    storage: &'a mut Storage<'b, T, D>,
+}
+
+impl<'a, 'b, T, D> OccupiedEntry<'a, 'b, T, D>
+    where T: Component,
+          D: Deref<Target = MaskedStorage<T>>,
+{
+    /// Get a reference to the component associated with the entity.
+    pub fn get(&self) -> Option<&T> {
+        self.storage.get(self.entity)
+    }
+}
+
+impl<'a, 'b, T, D> OccupiedEntry<'a, 'b, T, D>
+    where T: Component,
+          D: DerefMut<Target = MaskedStorage<T>>,
+{
+    /// Get a mutable reference to the component associated with the entity.
+    pub fn get_mut(&mut self) -> Option<&mut T> {
+        self.storage.get_mut(self.entity)
+    }
+
+    /// Inserts a value into the storage and returns the old one.
+    ///
+    /// Return value can be `None` if the entity passed in was dead.
+    pub fn insert(&mut self, component: T) -> Option<T> {
+        match self.storage.insert(self.entity, component) {
+            InsertResult::Updated(old) => Some(old),
+            _ => None,
+        }
+    }
+
+    /// Removes the component from the storage and returns it.
+    ///
+    /// Return value can be `None` if the entity passed in was dead.
+    pub fn remove(&mut self) -> Option<T> {
+        self.storage.remove(self.entity)
+    }
+}
+
+/// An entry to a storage which does not have a component associated to the entity.
+pub struct VacantEntry<'a, 'b: 'a, T: 'b, D: 'b> {
+    entity: Entity,
+    storage: &'a mut Storage<'b, T, D>,
+}
+
+impl<'a, 'b, T, D> VacantEntry<'a, 'b, T, D>
+    where T: Component,
+          D: DerefMut<Target = MaskedStorage<T>>,
+{
+    /// Inserts a value into the storage and returns the old one.
+    ///
+    /// Return value can be `None` if the entity passed in was dead.
+    pub fn insert(&mut self, component: T) -> Option<T> {
+        match self.storage.insert(self.entity, component) {
+            InsertResult::Updated(old) => Some(old),
+            _ => None,
+        }
+    }
+}
+
+/// Entry to a storage for convenient filling of components or removal based on whether
+/// the entity has a component.
+pub enum StorageEntry<'a, 'b: 'a, T: 'b, D: 'b> {
+    /// Entry variant that is returned if the entity does has a component.
+    Occupied(OccupiedEntry<'a, 'b, T, D>),
+    /// Entry variant that is returned if the entity does not have a component.
+    Vacant(VacantEntry<'a, 'b, T, D>),
+}
+
+pub enum EntryError {
```

Comment:
This belongs into the `error` module. Also, we should not have enums with just a single variant.